homeassistant/components/nest/__init__.py

```diff
@@ -134,40 +142,74 @@ def nest_update_event_broker(hass, nest):
         hass.async_create_task(hass.config_entries.async_forward_entry_setup(
             entry, component))
 
-    def set_mode(service):
-        """
-        Set the home/away mode for a Nest structure.
+    def cancel_eta(service):
+        """Cancel ETA for a Nest structure."""
+        if ATTR_STRUCTURE in service.data:
+            target_structures = service.data[ATTR_STRUCTURE]
+        else:
+            target_structures = hass.data[DATA_NEST].local_structure
+
+        for structure in nest.structures:
+            if structure.name in target_structures and structure.thermostats:
+                trip_id = service.data[ATTR_TRIP_ID]
+                _LOGGER.info("Cancelling ETA for trip: %s", trip_id)
+                structure.cancel_eta(trip_id)
+            else:
+                _LOGGER.error("Invalid structure: %s",
+                              service.data[ATTR_STRUCTURE])
 
-        You can set optional eta information when set mode to away.
-        """
+    def set_eta(service):
+        """Set mode to away and include ETA for a Nest structure."""
         if ATTR_STRUCTURE in service.data:
-            structures = service.data[ATTR_STRUCTURE]
+            target_structures = service.data[ATTR_STRUCTURE]
         else:
-            structures = hass.data[DATA_NEST].local_structure
+            target_structures = hass.data[DATA_NEST].local_structure
 
         for structure in nest.structures:
-            if structure.name in structures:
-                _LOGGER.info("Setting mode for %s", structure.name)
-                structure.away = service.data[ATTR_HOME_MODE]
+            if structure.name in target_structures and structure.thermostats:
+                _LOGGER.info("Setting mode for: %s to: %s",
+                             structure.name, HOME_MODE_AWAY)
+                structure.away = HOME_MODE_AWAY
+
+                now = datetime.utcnow()
+                trip_id = service.data.get(
+                    ATTR_TRIP_ID, "trip_{}".format(int(now.timestamp())))
+                eta_begin = now + service.data[ATTR_ETA]
+                eta_window = service.data.get(ATTR_ETA_WINDOW,
+                                              timedelta(minutes=1))
+                eta_end = eta_begin + eta_window
+                _LOGGER.info("Setting ETA for trip: %s, "
+                             "ETA window starts at: %s and ends at: %s",
+                             trip_id, eta_begin, eta_end)
+                structure.set_eta(trip_id, eta_begin, eta_end)
+            else:
+                _LOGGER.error("Invalid structure: %s",
+                              service.data[ATTR_STRUCTURE])
+
+    def set_mode(service):
```

Comment:
Is `set_away_mode` a better method name?

Reply:
I agree, I think `set_away_mode` is more intuitive. I would also suggest we change the parameter from `home_mode` to `mode` or `away_mode`.